### DISCUSSION TITLE
Chakra UI v3

### DISCUSSION THREAD

**Discussion by patrick-yendou** [csandman/chakra-react-select#319] [General]
Hey everyone,

V3 of ChakraUI is currently in the works. [Discussion](https://github.com/chakra-ui/chakra-ui/discussions/8518) and new [Docs](https://next.chakra-ui.com/). Any plans already to create a version of this package that supports v3 in the near future?

When I use the new release `TypeError: Cannot read properties of undefined (reading 'Input')`. Then I currently run into the following error:

```
TypeError: Cannot read properties of undefined (reading 'Input')
```

I assume it's because some of Chakra's components have changed.

**Comment by csandman** [comment]
Yeah I do plan on updating this for V3 when I get a chance! I have no idea how involved the migration path will be though, so I'll have to see when I have some time. I'm going to try and get started on it before v3 is fully released though.

**EDIT:** So I tried setting it up with V3 just to see, and I'm not even able to get to that point. I'm not even sure how to properly set up the `ChakraProvider` anymore haha. Do you have a basic example of how you set your project up?

**Reply by patrick-yendou** [reply]
What issue are you facing? This is the setup that works for me:

```
import {
  Box,
  Card,
  ChakraProvider,
  Text,
  defaultSystem,
} from "@chakra-ui/react";

export default function Home() {
  return (
    <ChakraProvider value={defaultSystem}>
      <Box></Box>
    </ChakraProvider>
  );
}
```

**Reply by csandman** [reply]
Ah, the `defaultSystem` is the part I couldn't find, I couldn't even find that documented in their new docs, but maybe I'm just missing something obvious 😅

**Reply by csandman** [reply]
Oh man, this is definitely going to end up being a major overhaul. I gave it a shot for a bit, and so many things have changed. I don't think this migration will be possible until they've completed the new docs for how the new theming works. Everything I've seen in the V3 docs for now are just the same as the old ones.

**Reply by wingsofovnia** [reply]
I did some work on migrating to v3 [here](https://github.com/csandman/chakra-react-select/compare/main...wingsofovnia:chakra-react-select:main). I am still confused with the new `useSlotRecipe` that replaced `useMultiStyleConfig`. Some other major changes that comes to mind:
* sx -> css
* size and variant props changed
* useTheme -> useChakraContext
* some Spinner attributes got removed

There are many tiny changes but overall all the components used here stayed intact. Most of their props are the same.

The [MIGRATION.md](https://github.com/chakra-ui/chakra-ui/blob/main/MIGRATION.md) is very helpful. The owner is pretty responsive as well and question get answered quickly [here](https://github.com/chakra-ui/chakra-ui/issues/8593).

I unfortunately gave up on migration as I lack understanding of both `useMultiStyleConfig` and `useSlotRecipe`. I wish I haven't jumped on Chakra 3.0.0-next 🥲 It feels like the Chakra release is not far away but it will take months until great libs like yours get on board. I hope it will get migrated soon and perhaps some work I've started could give you some hints.

**Reply by csandman** [reply]
Thanks for the head start at least! I also had some trouble figuring out the `useSlotRecipe` when I looked at it. I'll definitely use your branch as a jumping off point when I have time to fully tackle this though!

**Reply by csandman** [reply]
Also yeah, it probably was a bold move to hop on v3 before it's fully released 😅

**Comment by csandman** [comment]
Well, good news, I started working on getting v3 implemented last night (thanks for the starting point @wingsofovnia!) and it's looking pretty good so far! I ended up changing up where all of the built-in styles are being pulled from to use their new built-in select component, and it maps way better than it used to when I was pulling from the input and menu component styles. Here's a sneak peek of what it's looking like so far:

<img width="764" alt="image" src="https://github.com/user-attachments/assets/668ecfc6-fbd3-4265-94a4-1cfa7774742e">

<img width="771" alt="image" src="https://github.com/user-attachments/assets/717c8f73-dfa3-45b6-acef-1d5ca7af67f8">

I will probably have a `next` version out soon that people can try out and give feedback on.

**Reply by wingsofovnia** [reply]
Great work! Looking forward to replacing my in-house buggy thing with it :) Chakra UI is reportedly just a few weeks away from RC/final release.

Also, if you are looking towards mixing in "native" Chakra UI elements, you might be interested in Ark UI [Combobox](https://ark-ui.com/react/docs/components/combobox), which is more suitable for input + select scenarios. Ark UI is a unstyled components lib used by Chakra v3 and Chakra basically exposes Ark UI components with its styles. Unfortunately, Combobox itself isn't exposed _yet_ in Chakra but you can use this:

<details>
<summary>components/combobox/combobox.ts</summary>

```ts
import { Combobox as ArkCombobox } from '@ark-ui/react/combobox'
import {
  createSlotRecipeContext, defineSlotRecipe,
  type HTMLChakraProps,
  type SlotRecipeProps,
  type UnstyledProp,
} from '@chakra-ui/react'
import { comboboxAnatomy } from '@ark-ui/anatomy'

const comboboxSlotRecipe = defineSlotRecipe({
  className: 'combobox',
  slots: comboboxAnatomy.keys(),
  base: {
    root: {
      width: '100%',
    },
    label: {
      fontWeight: 'medium',
      color: 'gray.700',
      marginBottom: '2',
    },
    control: {
      display: 'flex',
      alignItems: 'center',
      width: '100%',
    },
    input: {
      width: '100%',
      padding: '2',
      borderRadius: 'md',
      border: '1px solid',
      borderColor: 'gray.200',
      _hover: {
        borderColor: 'gray.300',
      },
      _focus: {
        borderColor: 'blue.500',
        boxShadow: '0 0 0 1px var(--chakra-colors-blue-500)',
      },
    },
    trigger: {
      marginLeft: '2',
      padding: '2',
      borderRadius: 'md',
      backgroundColor: 'gray.100',
      _hover: {
        backgroundColor: 'gray.200',
      },
    },
    clearTrigger: {
      marginLeft: '2',
      padding: '2',
      borderRadius: 'md',
      color: 'gray.500',
      _hover: {
        color: 'gray.700',
      },
    },
    positioner: {
    },
    content: {
      zIndex: 'popover',
      backgroundColor: 'white',
      borderRadius: 'md',
      boxShadow: 'md',
      overflow: 'hidden',
    },
    itemGroup: {
      padding: '2',
    },
    itemGroupLabel: {
      fontWeight: 'bold',
      color: 'gray.600',
      fontSize: 'sm',
    },
    item: {
      display: 'flex',
      justifyContent: 'space-between',
      padding: '2',
      cursor: 'pointer',
      borderRadius: 'md',
      _hover: {
        backgroundColor: 'gray.100',
      },
      _selected: {
        backgroundColor: 'gray.100',
        color: 'black',
      },
    },
    itemText: {
    },
    itemIndicator: {
    },

  },
  variants: {},
  defaultVariants: {},
})

const { withProvider, withContext, useStyles: useComboboxStyles } = createSlotRecipeContext({ recipe: comboboxSlotRecipe })

export { useComboboxStyles }

export interface ComboboxRootProps
  extends HTMLChakraProps<'div', ArkCombobox.RootBaseProps<{}>>,
  SlotRecipeProps<'combobox'>,
  UnstyledProp {}

export const ComboboxRoot = withProvider<HTMLDivElement, ComboboxRootProps>(
  ArkCombobox.Root,
  'root',
  { forwardAsChild: true },
)

export interface ComboboxLabelProps extends HTMLChakraProps<'label', ArkCombobox.LabelBaseProps> {}

export const ComboboxLabel = withContext<HTMLLabelElement, ComboboxLabelProps>(
  ArkCombobox.Label,
  'label',
  { forwardAsChild: true },
)

export interface ComboboxControlProps
  extends HTMLChakraProps<'div', ArkCombobox.ControlBaseProps> {}

export const ComboboxControl = withContext<HTMLDivElement, ComboboxControlProps>(
  ArkCombobox.Control,
  'control',
  { forwardAsChild: true },
)

export interface ComboboxInputProps extends HTMLChakraProps<'input', ArkCombobox.InputBaseProps> {}

export const ComboboxInput = withContext<HTMLInputElement, ComboboxInputProps>(
  ArkCombobox.Input,
  'input',
  { forwardAsChild: true },
)

export interface ComboboxTriggerProps
  extends HTMLChakraProps<'button', ArkCombobox.TriggerBaseProps> {}

export const ComboboxTrigger = withContext<HTMLButtonElement, ComboboxTriggerProps>(
  ArkCombobox.Trigger,
  'trigger',
  { forwardAsChild: true },
)

export interface ComboboxClearTriggerProps
  extends HTMLChakraProps<'button', ArkCombobox.ClearTriggerBaseProps> {}

export const ComboboxClearTrigger = withContext<HTMLButtonElement, ComboboxClearTriggerProps>(
  ArkCombobox.ClearTrigger,
  'clearTrigger',
  { forwardAsChild: true },
)

export interface ComboboxPositionerProps
  extends HTMLChakraProps<'div', ArkCombobox.PositionerBaseProps> {}

export const ComboboxPositioner = withContext<HTMLDivElement, ComboboxPositionerProps>(
  ArkCombobox.Positioner,
  'positioner',
  { forwardAsChild: true },
)

export interface ComboboxContentProps
  extends HTMLChakraProps<'div', ArkCombobox.ContentBaseProps> {}

export const ComboboxContent = withContext<HTMLDivElement, ComboboxContentProps>(
  ArkCombobox.Content,
  'content',
  { forwardAsChild: true },
)

export interface ComboboxItemGroupProps
  extends HTMLChakraProps<'div', ArkCombobox.ItemGroupBaseProps> {}

export const ComboboxItemGroup = withContext<HTMLDivElement, ComboboxItemGroupProps>(
  ArkCombobox.ItemGroup,
  'itemGroup',
  { forwardAsChild: true },
)

export interface ComboboxItemGroupLabelProps
  extends HTMLChakraProps<'div', ArkCombobox.ItemGroupLabelBaseProps> {}

export const ComboboxItemGroupLabel = withContext<HTMLDivElement, ComboboxItemGroupLabelProps>(
  ArkCombobox.ItemGroupLabel,
  'itemGroupLabel',
  { forwardAsChild: true },
)

export interface ComboboxItemProps extends HTMLChakraProps<'div', ArkCombobox.ItemBaseProps> {}

export const ComboboxItem = withContext<HTMLDivElement, ComboboxItemProps>(
  ArkCombobox.Item,
  'item',
  { forwardAsChild: true },
)

export interface ComboboxItemTextProps
  extends HTMLChakraProps<'div', ArkCombobox.ItemTextBaseProps> {}

export const ComboboxItemText = withContext<HTMLDivElement, ComboboxItemTextProps>(
  ArkCombobox.ItemText,
  'itemText',
  { forwardAsChild: true },
)

export interface ComboboxItemIndicatorProps
  extends HTMLChakraProps<'div', ArkCombobox.ItemIndicatorBaseProps> {}

export const ComboboxItemIndicator = withContext<HTMLDivElement, ComboboxItemIndicatorProps>(
  ArkCombobox.ItemIndicator,
  'itemIndicator',
  { forwardAsChild: true },
)
```
</details>

**Reply by wingsofovnia** [reply]
Random: I was always puzzled why `padding-left` is bigger than `padding-y`. Seems like `next` will get the same:
<img width="249" alt="Знімок екрана 2024-10-10 о 21 01 28" src="https://github.com/user-attachments/assets/8afaa9e1-9f13-4ecf-a13b-71b2f3714acd">
Looks off to me or that's by design?

**Reply by csandman** [reply]
> Also, if you are looking towards mixing in "native" Chakra UI elements, you might be interested in Ark UI [Combobox](https://ark-ui.com/react/docs/components/combobox), which is more suitable for input + select scenarios. Ark UI is a unstyled components lib used by Chakra v3 and Chakra basically exposes Ark UI components with its styles. Unfortunately, Combobox itself isn't exposed _yet_ in Chakra but you can use this:

Yeah I was thinking that I could switch to something like that, but I feel like this package isn't really the right thing to do it. The whole purpose of this package was to fill the gap of the missing combobox originally, and if I end up just extending the native one, there isn't much purpose to it anymore haha, as people could just do the extending in their own projects. Plus the migration path for people using this package currently would be much larger than it already is.

Besides, `react-select` does offer a few more features that the native one doesn't, that people would otherwise have to implement themselves, such as the async and createable versions. Plus there are many other features customizable through simple props rather than implementing a solution from scratch. All that considered, I think I'll leave it to others to implement a more native solution if they want to.

---

> Random: I was always puzzled why `padding-left` is bigger than `padding-y`. Seems like `next` will get the same:
> <img width="249" alt="Знімок екрана 2024-10-10 о 21 01 28" src="https://github.com/user-attachments/assets/8afaa9e1-9f13-4ecf-a13b-71b2f3714acd">
> Looks off to me or that's by design?

Yeah, that padding is actually just pulling from the native padding on the new built-in select component haha (0.75rem). And the vertical padding isn't defined really, it's just a tag centered in the 2.5rem base height of the select. In Chakra v2, all of the input like components actually have a default horizontal padding of `0.75` rem, ~~but in Chakra v3, I noticed that the inputs actually have 0.5rem vs the 0.75rem of the select, so I'm honestly not really sure what to go with. I'm kind of leaning towards the 0.5 though, because you're right, it does look a bit off IMO. I think ultimately, I'm going to go with the `input` styles for the control field and the `select` styles for everything in the menu (and the indicators on the right), because this is really more of a combobox than a normal select, as you mentioned.~~ I realized I was wrong, they still have a `0.75rem` horizontal padding for inputs in v3, so I'll probably leave it as is.

My goal for this version is to define as few of my own internal style overrides as possible. In v2, I had to override a ton of the styles for things like the option (which I pull from the menu styles), but after my quick implementation last night, I realized that I can mostly leave it untouched. Whenever I do overrdies in the components themselves, it makes it much more difficult for the end user to override the styles just by modifying the theme styles for the component. If I can keep the overrides to an absolute minimum, ideally this would mean that any changes they make to their `select` theme would carry over to this component.

Also, all of the styles are subject to change, those screenshots were from a session of a few hours where I was just trying to see what I could do. It is by no means final. I'm thinking I'd like to get a very early version out asap though, for people who are trying to upgrade early, so at least people can help point out any issues I may have missed. There are so so many props for `react-select`, and different ways to customize it, so it's difficult for me to test it all on my own.

**Reply by csandman** [reply]
Also, for the horizontal padding, I think it looks more off with the multi-select because there is padding on the left side of the tag itself. In single mode, it looks pretty normal to me. In general, making the horizontal padding of components thicker than the vertical is a pretty common UI pattern.

Because of the tag thing though, I probably will end up pulling the control styles from the input.

**Reply by wingsofovnia** [reply]
Agreed on all points 🙂 Looking forward to giving `next` version a try as soon as it's released.

upd: I just realized that my illustration compares padding to the _label height_ not _padding-y_ 🤦‍♂️ I'm glad you got the idea anyway 😀

upd2: I was redoing the illustration for no reason and while duplicating padding-top measurement, I noticed padding-bottom is just a hair smaller:
<img width="342" alt="Знімок екрана 2024-10-11 о 09 13 04" src="https://github.com/user-attachments/assets/ffc0994a-7c35-4e11-bfec-5b88f98772eb">


This is a in the pure "pixel-perfect" area and might pretty much be a distortion in the Preview editor or smth but I decided to share this anyway just fyi.

**Reply by csandman** [reply]
> upd: I just realized that my illustration compares padding to the _label height_ not _padding-y_ 🤦‍♂️ I'm glad you got the idea anyway 😀

Haha yeah I did notice that, but your description made it pretty clear. Plus, it helped that I've noticed the same thing in the past.

Also, I realized that i was wrong, and the input actually does have the `0.75rem` horizontal padding, same as the select. Because of that, I think I'm actually going to leave it up to the end user to decide how to handle it. You can always change it to `0.5rem` in the select/input themes!



> upd2: I was redoing the illustration for no reason and while duplicating padding-top measurement, I noticed padding-bottom is just a hair smaller:

And yeah, that's probably due to the tag being a non-pixel perfect height. I checked the inspector and it looks like it has a height of `24.8px`. Not that it matters that much though haha, it's not particularly noticeable.

<img width="446" alt="image" src="https://github.com/user-attachments/assets/ccf46308-1f31-449a-a35f-e04a48a80a9d">

**Comment by csandman** [comment]
Ok I published my first draft of v6 (Chakra v3)! It's still very much a work in progress, so I would expect it to change, but I wanted to get eyes on it as soon as possible. If you want to check it out, you can see the draft PR I put up here: https://github.com/csandman/chakra-react-select/pull/341

Or install it using the `next` tag:

```sh
npm i chakra-react-select@next
# or
yarn add chakra-react-select@next
```

If you have any issues with it, leave a comment on my draft PR (or a review), or post something here. I'd rather not clog up the normal issues with things about this very much not ready draft haha.

There is also no documentation of changes yet. There are many, so that will take a while, but I want to finalize them before I go through with writing updated docs.

**Reply by csandman** [reply]
I just pushed a new `v6.0.0-next.2` version that is now officially made for the v3 release of Chakra for anyone who wants to test it! I'm still not fully confident that it's ready to go, but it's definitely pretty close, so if you're interested in upgrading to Chakra v3, give it a try and let me know!

**Comment by csandman** [comment]
Good news, I made some great progress on Chakra React Select v6, and have released the first release candidate version. You can install it with the following:

```sh
npm i chakra-react-select@rc
```

You can read more about the most recent changes in this comment: https://github.com/csandman/chakra-react-select/pull/341#issuecomment-2516244067

**Comment by lucas-garrido** [comment]
react 19 is now stable could you add it as peer dep please ?

**Reply by csandman** [reply]
Yeah, good call, I had been thinking about that. I'll do a new RC release later today with an updated dependency range.

**Reply by csandman** [reply]
Ok, I just pushed a new version `6.0.0-rc.2` which expands the react peer dep range to `>=18` to match that of `@chakra-ui/react@3`. One important thing to note is that [`react-select@5.8.3`](https://github.com/JedWatson/react-select/releases/tag/react-select%405.8.3) (the most recent version of it) implemented some type compatibility with React 19, so you may need to run `npm upgrade` or something to ensure your version of `react-select` is up to date. Another way to potentially do so if it's out of date is to uninstall this package entirely and do a fresh install.

**Comment by csandman** [comment]
Ok all, the official release is most likely coming Monday! Thanks for all the support!

**Comment by maulik13** [comment]
Just a question, I am using RC.4 and wondering how to allow more variants for type `Variant`? It is not picking up my customization to the select recipe even with typegen. 

I am using `variant='filled'` and it works fine, but TS compiler gives the error. 

Reference.
https://github.com/csandman/chakra-react-select/blob/a965da5ecd16be43d7a3f9bcece21f11e8dc7fdb/src/types.ts#L45

**Reply by csandman** [reply]
Ah, good to know. Hmm, can you show me how you extended the theme so I can have an example to work with, after doing a type gen? I assume the types work properly for you with the normal select?

**Reply by csandman** [reply]
Actually, I got an example working with the demo project. I switched the typing on the `variant` prop to be `SelectRootProps["variant"]`, and that appears to work with the statically generated theme props! I just pushed a new version with this change, `v6.0.0-rc.5`.

**Reply by maulik13** [reply]
It works with 6.0.0!!! You are a champ :)

**Comment by csandman** [comment]
v6 is officially released, give it a try and let me know if anything isn't working right!
